Wallet standard interface proposal (#1478)
stellar/stellar-protocol: ecosystem/sep-0043.md

## Specification

The standard web wallet interface provides a structure for common methods that wallets are expected to offer. In addition, the interface defines common error message shapes to ensure wallets adhere to informative and usable error messaging. Messages are designed to be surfaced directly to the user, so they should provide a readable description of what went wrong.

### Wallet Interface Format

Wallets will employ an API using the following format:

```typescript

interface Error {
  message: string,    // general description message returned to the client app
  code: number,       // unique error code
  ext?: Array<string>  // optional extended details
}

{
  getAddress: () => Promise<{address: string;} | Error>,
  signTransaction: (xdr: string, opts?: { network?: string, networkPassphrase?: string, address?: string }) => Promise<{signedXdr: string} | Error>,
  signAuthEntry: (authEntryXdr: string, opts?: { network?: string, networkPassphrase?: string, address?: string }) => Promise<{signedAuthEntry: string} | Error>,

>   signMessage: (message: string, opts?: { network?: string, networkPassphrase?: string, address?: string }) => Promise<{signedMessage: string} | Error>,

Reviewer: :bulb: For all of these, the response should just be `Promise<{ signature: string } | Error>`. It's a better reflection of the return value, which is just an opaque signature. Also, we should describe its format: is it hex, base64, etc.? For example, [`Keypair.sign`](https://stellar.github.io/js-stellar-sdk/Keypair.html#sign) just returns raw bytes.

Unless this is implying that you would get e.g. an auth entry that is "filled out" with the signature field inside of it? This isn't explicit in the description below.

Author: @Shaptic This is a good call out for `signAuthEntry` and `signMessage`. 

`signTransaction` actually returns the signed transaction xdr from Freighter, not just the signautre, so I do think `Promise<{ xdr: string } | Error>` is still correct here. Though it may be useful to also pass just the signature

Author: Good question on the Buffer encoding piece. I know Albedo encodes as hex. I'll check with the ecosystem to see if they have a preference here

### Wallet Interface Format

Wallets will employ an API using the following format:

```typescript

interface Error {
  message: string,    // general description message returned to the client app
  code: number,       // unique error code
  ext?: Array<string>  // optional extended details
}

{
  get_address: () => Promise<{address: string;} | Error>,
  
  sign_transaction: (xdr: string, opts?: { network_passphrase?: string, address?: string; submit?: boolean; submit_url?: string; }) => Promise<{signed_tx_xdr: string; signer_address: string;} | Error>,

  sign_auth_entry: (auth_entry: string, opts?: { network_passphrase?: string, address?: string }) => Promise<{signed_auth_entry: string; signer_address: string;} | Error>,

  sign_message: (message: string, opts?: { network_passphrase?: string, address?: string }) => Promise<{signed_message: string; signer_address: string;} | Error>,

>   get_network: () => Promise<{network: string, network_url: string, network_passphrase: string, soroban_rpc_url: string}>

Author: Ah, that's a good point. Dapps should be able to target their own Horizon/RPC url's based on the network and what the wallet is configured to is actually kind of irrelevant. I'll change to just `network/network_passphrase`

## Preamble

```
SEP: 0043
Title: Standard Web Wallet API Interface
Authors: Piyal Basu <piyal@stellar.org>, Leigh McCulloch <@leighmcculloch>, George Kudrayvtsev <george@stellar.org>, Enrique Arrieta <@earrietadev>, Orbit Lens <@orbitlens>
Track: Standard
Status: Draft
Created: 2024-04-11
Version: 1.1.0
Discussion: https://github.com/stellar/stellar-protocol/discussions/1467
```

## Abstract

This standard aims to enhance developer experiences by standardizing an interface for web wallet's to use
when interacting with other codebases.

## Motivation

There are numerous Stellar wallets in the ecosystem with this number expected to grow as the ecosystem does.

> Wallets expose an API to allow dapps to call methods that preform such common tasks as transmitting a wallet's adress. An issue is that wallets all use slightly different API's to preform similar functions, with different method names and call signature as well as different return values. Developers have a motivation to integrate with as many as possible, but currently, they must create abstractions to handle every wallet they wish to connect to. This proposal presents a standard API interface for all wallets to implement. If all wallets have the same interface, it removes the needs for developers to create abstractions for every individual wallet they want to integrate with. Developers can safely assume a wallet uses this agreed upon interface, creating simpler codepaths that easily connect to current wallets while also simplifying the onboarding of new wallets.

Reviewer 2: ```suggestion
Wallets expose an API to allow dapps to call methods that preform such common tasks as transmitting a wallet's address. An issue is that wallets all use slightly different API's to preform similar functions, with different method names and call signature as well as different return values. Developers have a motivation to integrate with as many as possible, but currently, they must create abstractions to handle every wallet they wish to connect to. This proposal presents a standard API interface for all wallets to implement. If all wallets have the same interface, it removes the needs for developers to create abstractions for every individual wallet they want to integrate with. Developers can safely assume a wallet uses this agreed upon interface, creating simpler codepaths that easily connect to current wallets while also simplifying the onboarding of new wallets.
```

## Specification

The standard web wallet interface provides a structure for common methods that wallets are expected to offer. In addition, the interface defines common error message shapes to ensure wallets adhere to informative and usable error messaging. Messages are designed to be surfaced directly to the user, so they should provide a readable description of what went wrong.

### Wallet Interface Format

Wallets will employ an API using the following format:

```typescript

interface Error {
  message: string,    // general description message returned to the client app
  code: number,       // unique error code
  ext?: Array<string>  // optional extended details
}

{
  get_address: () => Promise<{address: string;} | Error>,

>   sign_transaction: (xdr: string, opts?: { network_passphrase?: string, address?: string; submit?: boolean; submit_url?: string; }) => Promise<{signed_tx_xdr: string; signer_address: string;} | Error>,

Reviewer 2: ```suggestion
  sign_transaction: (xdr: string; opts?: { network_passphrase?: string; address?: string; submit?: boolean; submit_url?: string; }) => Promise<{signed_tx_xdr: string; signer_address: string;} | Error>,
```

  sign_message: (message: string, opts?: { network_passphrase?: string, address?: string }) => Promise<{signed_message: string; signer_address: string;} | Error>,

  get_network: () => Promise<{network: string, network_passphrase: string}>
}
```

In addition, this proposal has a standard for error interfaces. Below are examples of different kinds of errors a wallet may surface:

```typescript
{
  unhandled: { // internal wallet error, likely caused by the wallet logic itself
    message: 'Unhandled error occurred.',
    code: -1
  },
  external: { // external service (Horizon, RPC, etc.) returned an error 
    message: 'External error occurred.',
    code: -2,
    ext: ['Operation 2: Invalid "amount" value.', "Operation 5: Asset issuer is required."] // malformed tx error example
  },
  invalid: { // client app request is invalid (wrong parameters, invalid transaction XDR, etc.)
    message: 'Request is invalid.',
    code: -3,
    ext: ['Invalid transaction XDR'] // example error
  },
  rejected: { // user rejected the request, client app should not try to retry the request

>     message: 'Rejected by the user.',

Reviewer 2: There is also the localization related to user-facing messages, ideally they should appear on a given language based on where user is located or based on some user config. Would this be part of the scope of this SEP?

>     message: 'Rejected by the user.',

Author: This is a good callout. I can make this more clear - `message` should provide users an action that can take to remedy the situation. The contents of the `message` will be defined by the wallet itself. We're really just asking them to group their errors by `unhandled|external|invalid|rejected`.

>     message: 'Rejected by the user.',

Author: I think localization is a good callout, as well. A wallet should be localizing the `message` and `ext` field and sending it back. I'll add that to the SEP

>     message: 'Rejected by the user.',

Reviewer 2: Looks good, thanks for the changes!

### sign_message

Similar to its transaction and aither entry counterparts, `sign_message` will take an arbitrary message provided by the requester and apply a signature to it before returning a signed message and the signer address back to the requester. This can be used to implement identity/ownership verification. The first parameter is the XDR string while the second parameter is a set of options. Options allow a user to specify what network they intend the transaction to be signed on. They can do so by passing the network's passphrase as `network_passphrase`. Options also allow a wallet to specify which account they're seeking a signature for as `address`. If a wallet holds numerous addresses, it can use this param to ensure it is signing with the intended address.

### get_network

`get_network` provides details about the network that the wallet is currently configured to. This allows a dapp to confirm that they are building a transaction for the correct network.

### Errors

> Error messaging provides clear errors to the requester. The `message` field provides an informative, human readable string that a UI could surface to the end user. The 4 types of `code` key dilineate between possible erroes: those caused by an internal wallet error, an error caused by external service (for ex: Horizon or Soroban RPC), invalid user input (for ex: a malformed XDR), and an error caused by the user rejecting the action requested. Finally, the `ext` field is an optional value that can provide more context for the error, if needed.

Reviewer 2: ```suggestion
Error messaging provides clear errors to the requester. The `message` field provides an informative, human readable string that a UI could surface to the end user. The 4 types of `code` key delineate between possible errors: those caused by an internal wallet error, an error caused by external service (for ex: Horizon or Soroban RPC), invalid user input (for ex: a malformed XDR), and an error caused by the user rejecting the action requested. Finally, the `ext` field is an optional value that can provide more context for the error, if needed.
```

They can do so by passing the network's passphrase as `network_passphrase`. Options also allow a wallet to specify which
account they're seeking a signature for as `address`. If a wallet holds numerous addresses, it can use this param to
ensure it is signing with the intended address.

### get_network

`get_network` provides details about the network that the wallet is currently configured to. This allows a dapp to
confirm that they are building a transaction for the correct network.

### Errors

Error messaging provides clear errors to the requester. The `message` field provides an informative, human readable
string that a UI could surface to the end user. This `message` should include actions a user can take to remedy the
situation. The 4 types of `code` key delineate between possible errors: those caused by an internal wallet error, an
error caused by external service (for ex: Horizon or Soroban RPC), invalid user input (for ex: a malformed XDR), and an
error caused by the user rejecting the action requested. Finally, the `ext` field is an optional value that can provide

> more context for the error, if needed. The wallet is reponsible for localizing the `message` and `ext` fields to be

Reviewer 2: ```suggestion
more context for the error, if needed. The wallet is responsible for localizing the `message` and `ext` fields to be
```

## Preamble

```
SEP: 0043
Title: Standard Web Wallet API Interface
Authors: Piyal Basu <piyal@stellar.org>, Leigh McCulloch <@leighmcculloch>, George Kudrayvtsev <george@stellar.org>, Enrique Arrieta <@earrietadev>, Orbit Lens <@orbitlens>
Track: Standard
Status: Draft
Created: 2024-04-11
Version: 1.2.0
Discussion: https://github.com/stellar/stellar-protocol/discussions/1467
```

## Abstract

This standard aims to enhance developer experiences by standardizing an interface for web wallet's to use when
interacting with other codebases.

## Motivation

There are numerous Stellar wallets in the ecosystem with this number expected to grow as the ecosystem does. Wallets

> expose an API to allow dapps to call methods that preform such common tasks as transmitting a wallet's address. An issue

Reviewer: nit: typo
```suggestion
expose an API to allow dapps to call methods that perform common tasks such as transmitting a wallet's address. An issue
```

> is that wallets all use slightly different API's to preform similar functions, with different method names and call

Reviewer: ```suggestion
is that wallets all use slightly different API's to perform similar functions, with different method names and call
```

> signature as well as different return values. Developers have a motivation to integrate with as many as possible, but

Reviewer: ```suggestion
signatures as well as different return values. Developers have a motivation to integrate with as many as possible, but
```

Author: Thanks - the lack of spell check in vscode is very evident in this PR 😬

standard API interface for all wallets to implement. If all wallets have the same interface, it removes the needs for
developers to create abstractions for every individual wallet they want to integrate with. Developers can safely assume
a wallet uses this agreed upon interface, creating simpler codepaths that easily connect to current wallets while also
simplifying the onboarding of new wallets.

## Specification

The standard web wallet interface provides a structure for common methods that wallets are expected to offer. In
addition, the interface defines common error message shapes to ensure wallets adhere to informative and usable error
messaging. Messages are designed to be surfaced directly to the user, so they should provide a readable description of
what went wrong.

### Wallet Interface Format

Wallets will employ an API using the following format:

```typescript

interface Error {
  message: string,    // general description message returned to the client app
  code: number,       // unique error code
  ext?: Array<string>  // optional extended details
}

{
  getAddress: () => Promise<{address: string;} | Error>,

>   signTransaction: (xdr: string, opts?: { networkPassphrase?: string, address?: string; submit?: boolean; submitUrl?: string; }) => Promise<{signedTxXdr: string; signerAddress: string;} | Error>,

Reviewer: Can we wrap these so they're easier to read in the rendered markdown?

  signAuthEntry: (authEntry: string, opts?: { networkPassphrase?: string, address?: string }) => Promise<{signedAuthEntry: string; signerAddress: string;} | Error>,

  signMessage: (message: string, opts?: { networkPassphrase?: string, address?: string }) => Promise<{signedMessage: string; signerAddress: string;} | Error>,

  getNetwork: () => Promise<{network: string, networkPassphrase: string}>
}
```

In addition, this proposal has a standard for error interfaces. Below are examples of different kinds of errors a wallet

> may surface:

Reviewer: We should heavily emphasize which parts of the errors are standard (i.e. the interface or categories) vs. examples that wallets should fill in themselves (i.e. the messages/codes themselves, mentioning localization etc. might be worthwhile), because it's not super clear to me here which ones are "guaranteed" for an implementor.

```typescript
{
  unhandled: { // internal wallet error, likely caused by the wallet logic itself
    message: 'The wallet encountered an internal error. Please try again or contact the wallet if the problem persists.',
    code: -1
  },
  external: { // external service (Horizon, RPC, etc.) returned an error
    message: 'An error occurred with Horizon. Please try again.',
    code: -2,
    ext: ['Operation 2: Invalid "amount" value.', "Operation 5: Asset issuer is required."] // malformed tx error example
  },
  invalid: { // client app request is invalid (wrong parameters, invalid transaction XDR, etc.)
    message: 'Request is invalid. Please check the details and try again.',
    code: -3,
    ext: ['Invalid transaction XDR'] // example error
  },
  rejected: { // user rejected the request, client app should not try to retry the request
    message: 'The user rejected this request.',
    code: -4
  }
}

```

### Methods

### getAddress

> `getAddress` will provide the public key the wallet is signing for. This may be the account loaded in the wallet, or it

Reviewer: Always `G...`, right?

Author: Yes that's correct. I'll specify

may be a different account that the wallet is the signer for.

### signTransaction

> `signTransaction` will take an XDR provided by the requester and apply a signature to it before returning a signed XDR
> and the signer address back to the requester. The first parameter is the XDR string while the second parameter is a set

Reviewer: Suggested rephrase to provide more detail, somewhat applies to everywhere you mention XDR strings

```suggestion
`signTransaction` will take a serialized transaction provided by the requester and apply a signature to it before returning a signed XDR and the signer address back to the requester. The first parameter is a base64-encoded string of an XDR `Transaction` instance (right? or is it the envelope? plus cross-reference to the `.x` file/line itself) while the second parameter is a set
```

Also, does

> signed XDR

mean a `TransactionEnvelope` with a corresponding `DecoratedSignature` appended?

seeking a signature for as `address`. If a wallet holds numerous addresses, it can use this param to ensure it is
signing with the intended address. Options also allow a user to set a `submit` flag, which will allow a user to sign a
transaction and immediately submit it to a network. If a user would like to configure the URL of the network to submit
to, they can do so by setting `submitUrl`.

### signAuthEntry

Similar to its transaction counterpart, `signAuthEntry` will take an XDR provided by the requester and apply a signature
to it before returning a signature and the signer address back to the requester. The first parameter is the XDR string
while the second parameter is a set of options. Options allow a user to specify what network they intend the transaction
to be signed on. They can do so by passing the network's passphrase as `networkPassphrase`. Options also allow a wallet
to specify which account they're seeking a signature for as `address`. If a wallet holds numerous addresses, it can use
this param to ensure it is signing with the intended address.

### signMessage

> Similar to its transaction and aither entry counterparts, `signMessage` will take an arbitrary message provided by the

Reviewer: typo:
```suggestion
Similar to its transaction and auth entry counterparts, `signMessage` will take an arbitrary message provided by the
```

also, wouldn't it be `Contrary` rather than `Similar` given that it's now arbitrary rather than well-defined or structured data?

Author: I was trying to convey that all 3 are similar in that you're sending _something_ to the wallet, it signs, and then returns the signed version. But I can make that more clear

    message: 'An error occurred with Horizon. Please try again.',
    code: -2,
    ext: ['Operation 2: Invalid "amount" value.', "Operation 5: Asset issuer is required."] // malformed tx error example
  },

  // Either the dapp or the user passed a malformed request to the wallet. For example, a user didn't pass a valid XDR to `signTransaction`

  {
    message: 'Request is invalid. Please check the details and try again.',
    code: -3,
    ext: ['Invalid transaction XDR'] // example error
  },

  // The wallet received the request from the client, but the user decided to decline.

  rejected: {
    message: 'The user rejected this request.',
    code: -4
  }

```

### Methods

### getAddress

`getAddress` will provide the public key (G...) the wallet is signing for. This may be the account loaded in the wallet,
or it may be a different account that the wallet is the signer for.

### signTransaction

`signTransaction` will take an XDR provided by the requester and apply a signature to it before returning a
base64-encoded string XDR-encoded
[Transaction Envelope](https://github.com/stellar/js-stellar-base/blob/5efdff0b2a78ed842f7463340b4ff23eb4370635/types/next.d.ts#L11928)
with

> [Decorated Signatures](https://github.com/stellar/js-stellar-base/blob/5efdff0b2a78ed842f7463340b4ff23eb4370635/types/next.d.ts#L5882)

Reviewer: ```suggestion
[Decorated Signatures](https://github.com/stellar/js-stellar-base/blob/5efdff0b2a78ed842f7463340b4ff23eb4370635/types/curr.d.ts#L5882)
```

(They're identical today, but referencing `curr` will be more accurate in the future once `next` changes.)
